src/mesh/mesh_optimizer.hh

```diff
@@ -0,0 +1,437 @@
+
+#ifndef MESH_OPTIMIZER_HH_
+#define MESH_OPTIMIZER_HH_
+
+#include <vector>
+
+#include <armadillo>
+
+#include "mesh/mesh.h"
+#include "mesh/accessors.hh"
+#include "mesh/range_wrapper.hh"
+#include "bounding_box.hh"
+
+struct Vec3 {
+    inline Vec3(arma::vec3 other) {
+        for (uint i = 0; i < 3; ++i) {
+            data[i] = other[i];
+        }
+    }
+    inline arma::vec3 arma() const {
+        return data;
+    }
+    inline double& operator[](uint i) {
+        return data[i];
+    }
+    inline double operator[](uint i) const {
+        return data[i];
+    }
+    double data[3];
+};
+
+struct Permutee {
+    inline Permutee(uint original_index, double curve_value) : original_index_(original_index), curve_value_(curve_value) {}
+    uint original_index_;
+    double curve_value_;
+};
+
+struct Normalizer {
```

Comment:
This should perform mapping of the mesh bounding box to the unit cube. 
Make it simple class with overloaded call operator.